### PR TITLE
Duplictae policy

### PR DESCRIPTION
**Pull request by AvitalFineRedis**
add support for DUPLICATE_POLICY on TS.CREATE and TS.ADD #37

### CODE REVIEW

File: NRedisTimeSeries/Commands/DuplicatePolicy.cs [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+        /// </summary>
+        MAX,
+
+        /// <summary>
+        /// If a previous sample exists, add the new sample to it so that the updated value is equal to (previous + new).
+        /// If no previous sample exists, set the updated value equal to the new value.
+        /// </summary>
+        SUM
+    }
+}
```
Comment (DvirDukhan) [inline NRedisTimeSeries/Commands/DuplicatePolicy.cs:41]: add an empty line.

File: NRedisTimeSeries/Commands/DuplicatePolicy.cs [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+        /// </summary>
+        LAST,
+
+        /// <summary>
+        /// only override if the value is lower than the existing value
+        /// </summary>
+        MIN,
+
+        /// <summary>
+        /// only override if the value is higher than the existing value
```
Comment (DvirDukhan) [inline NRedisTimeSeries/Commands/DuplicatePolicy.cs:31]: Start comment with a capital letter, end with a period.

File: NRedisTimeSeries/Commands/DuplicatePolicy.cs [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+        /// </summary>
+        FIRST,
+
+        /// <summary>
+        /// override with latest value
+        /// </summary>
+        LAST,
+
+        /// <summary>
+        /// only override if the value is lower than the existing value
```
Comment (DvirDukhan) [inline NRedisTimeSeries/Commands/DuplicatePolicy.cs:26]: Start comment with a capital letter, end with a period.

File: NRedisTimeSeries/Commands/DuplicatePolicy.cs [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+        /// </summary>
+        BLOCK,
+
+        /// <summary>
+        /// ignore the new value
+        /// </summary>
+        FIRST,
+
+        /// <summary>
+        /// override with latest value
```
Comment (DvirDukhan) [inline NRedisTimeSeries/Commands/DuplicatePolicy.cs:21]: Start comment with a capital letter, end with a period.

File: NRedisTimeSeries/Commands/DuplicatePolicy.cs [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+    /// </summary>
+    public enum TsDuplicatePolicy
+    {
+        /// <summary>
+        /// an error will occur for any out of order sample
+        /// </summary>
+        BLOCK,
+
+        /// <summary>
+        /// ignore the new value
```
Comment (DvirDukhan) [inline NRedisTimeSeries/Commands/DuplicatePolicy.cs:16]: Start comment with a capital letter, end with a period.

File: NRedisTimeSeries/Commands/DuplicatePolicy.cs [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+namespace NRedisTimeSeries.Commands
+{ 
+    /// <summary>
+    /// Policy to handle duplicate samples. 
+    /// The default policy for database-wide is BLOCK
+    /// </summary>
+    public enum TsDuplicatePolicy
+    {
+        /// <summary>
+        /// an error will occur for any out of order sample
```
Comment (DvirDukhan) [inline NRedisTimeSeries/Commands/DuplicatePolicy.cs:11]: Start comment with a capital letter, end with a period.

File: NRedisTimeSeries/DataTypes/TimeSeriesInformation.cs [hunk 7]
```diff
@@ -65,7 +66,9 @@ public class TimeSeriesInformation
         /// </summary>
         public IReadOnlyList<TimeSeriesRule> Rules { get; private set; }
 
-        internal TimeSeriesInformation(long totalSamples, long memoryUsage, TimeStamp firstTimeStamp, TimeStamp lastTimeStamp, long retentionTime, long chunkCount, long chunkSize, IReadOnlyList<TimeSeriesLabel> labels, string sourceKey, IReadOnlyList<TimeSeriesRule> rules)
+        public TsDuplicatePolicy? DuplicatePolicy {  get; private set; }
```
Comment (DvirDukhan) [inline NRedisTimeSeries/DataTypes/TimeSeriesInformation.cs:69]: why optional and no string?

File: NRedisTimeSeries.Test/TestAPI/TestAddAsync.cs [hunk 8]
```diff
@@ -107,6 +108,90 @@ public async Task TestAddWithUncompressed()
...
+
+        [Fact]
+        public async Task TestAddWithDuplicatePolicyMin()
+        { 
+            var key = CreateKeyName();
+            var db = redisFixture.Redis.GetDatabase();
+            TimeStamp timeStamp = DateTime.UtcNow;
+            Assert.Equal(timeStamp, await db.TimeSeriesAddAsync(key, timeStamp, 1.1));
+
+            // Insert a bigger number and check that it did not change the value
```
Comment (DvirDukhan) [inline NRedisTimeSeries.Test/TestAPI/TestAddAsync.cs:131]:
```suggestion
            // Insert a bigger number and check that it did not change the value.
```

File: NRedisTimeSeries.Test/TestAPI/TestAddAsync.cs [hunk 9]
```diff
@@ -107,6 +108,90 @@ public async Task TestAddWithUncompressed()
...
+            Assert.Equal(timeStamp, await db.TimeSeriesAddAsync(key, timeStamp, 1.2, policy: TsDuplicatePolicy.MIN));
+            IReadOnlyList<TimeSeriesTuple> results = await db.TimeSeriesRangeAsync(key, timeStamp, timeStamp);
+            Assert.Equal(1.1, results[0].Val);
+
+            // Insert a smaller number and check that it changed
```
Comment (DvirDukhan) [inline NRedisTimeSeries.Test/TestAPI/TestAddAsync.cs:136]:
```suggestion
            // Insert a smaller number and check that it changed.
```

File: NRedisTimeSeries.Test/TestAPI/TestAddAsync.cs [hunk 10]
```diff
@@ -107,6 +108,90 @@ public async Task TestAddWithUncompressed()
...
+
+        [Fact]
+        public async Task TestAddWithDuplicatePolicyMax()
+        { 
+            var key = CreateKeyName();
+            var db = redisFixture.Redis.GetDatabase();
+            TimeStamp timeStamp = DateTime.UtcNow;
+            Assert.Equal(timeStamp, await db.TimeSeriesAddAsync(key, timeStamp, 1.1));
+
+            // Insert a smaller number and check that it did not change the value
```
Comment (DvirDukhan) [inline NRedisTimeSeries.Test/TestAPI/TestAddAsync.cs:150]:
```suggestion
            // Insert a smaller number and check that it did not change the value.
```

File: NRedisTimeSeries.Test/TestAPI/TestAddAsync.cs [hunk 11]
```diff
@@ -107,6 +108,90 @@ public async Task TestAddWithUncompressed()
...
+            Assert.Equal(timeStamp, await db.TimeSeriesAddAsync(key, timeStamp, 1.0, policy: TsDuplicatePolicy.MAX));
+            IReadOnlyList<TimeSeriesTuple> results = await db.TimeSeriesRangeAsync(key, timeStamp, timeStamp);
+            Assert.Equal(1.1, results[0].Val);
+            // Insert a bigger number and check that it changed
```
Comment (DvirDukhan) [inline NRedisTimeSeries.Test/TestAPI/TestAddAsync.cs:154]:
```suggestion
            // Insert a bigger number and check that it changed.
```

File: NRedisTimeSeries.Test/TestAPI/TestAdd.cs [hunk 12]
```diff
@@ -102,6 +103,76 @@ public void TestAddWithChunkSize()
...
+        }
+
+        [Fact]
+        public void TestAddWithDuplicatePolicyMin()
+        { 
+            IDatabase db = redisFixture.Redis.GetDatabase();
+            TimeStamp now = DateTime.UtcNow;
+            Assert.Equal(now, db.TimeSeriesAdd(key, now, 1.1));
+
+            // Insert a bigger number and check that it did not change the value
```
Comment (DvirDukhan) [inline NRedisTimeSeries.Test/TestAPI/TestAdd.cs:124]:
```suggestion
            // Insert a bigger number and check that it did not change the value.
```

File: NRedisTimeSeries.Test/TestAPI/TestAdd.cs [hunk 13]
```diff
@@ -102,6 +103,76 @@ public void TestAddWithChunkSize()
...
+            Assert.Equal(now, db.TimeSeriesAdd(key, now, 1.2, policy: TsDuplicatePolicy.MIN));
+            Assert.Equal(1.1, db.TimeSeriesRange(key, now, now)[0].Val);
+            // Insert a smaller number and check that it changed
```
Comment (DvirDukhan) [inline NRedisTimeSeries.Test/TestAPI/TestAdd.cs:127]:
```suggestion
            // Insert a smaller number and check that it changed.
```

File: NRedisTimeSeries.Test/TestAPI/TestAdd.cs [hunk 14]
```diff
@@ -102,6 +103,76 @@ public void TestAddWithChunkSize()
...
+        }
+
+        [Fact]
+        public void TestAddWithDuplicatePolicyMax()
+        { 
+            IDatabase db = redisFixture.Redis.GetDatabase();
+            TimeStamp now = DateTime.UtcNow;
+            Assert.Equal(now, db.TimeSeriesAdd(key, now, 1.1));
+
+            // Insert a smaller number and check that it did not change the value
```
Comment (DvirDukhan) [inline NRedisTimeSeries.Test/TestAPI/TestAdd.cs:139]:
```suggestion
            // Insert a smaller number and check that it did not change the value.
```

File: NRedisTimeSeries.Test/TestAPI/TestAdd.cs [hunk 15]
```diff
@@ -102,6 +103,76 @@ public void TestAddWithChunkSize()
...
+            Assert.Equal(now, db.TimeSeriesAdd(key, now, 1.0, policy: TsDuplicatePolicy.MAX));
+            Assert.Equal(1.1, db.TimeSeriesRange(key, now, now)[0].Val);
+            // Insert a bigger number and check that it changed
```
Comment (DvirDukhan) [inline NRedisTimeSeries.Test/TestAPI/TestAdd.cs:142]:
```suggestion
            // Insert a bigger number and check that it changed.
```

**Review by DvirDukhan — COMMENTED** [review]
Looks good
few comments

File: NRedisTimeSeries/DataTypes/TimeSeriesInformation.cs (hunk 7 above)
Comment (AvitalFineRedis) [reply NRedisTimeSeries/DataTypes/TimeSeriesInformation.cs:69]: Why string if we have enum for that purpose?

**Review by DvirDukhan — APPROVED** [review]

**Pull request merged**